File: C_plus_plus/Path_Sum_II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        if(!root) return result;
        dsp(root,vector<int>(),0,sum);
        return result;
    }
    void dsp(TreeNode *node , vector<int> prev , int pre , int sum){
        if(node->left==NULL&&node->right==NULL&&pre+node->val==sum) {
            prev.push_back(node->val);
            result.push_back(prev);
            return;
        }
        prev.push_back(node->val);
        if(node->left) dsp(node->left,prev,pre+node->val,sum);
        if(node->right) dsp(node->right,prev,pre+node->val,sum);
    }
};

File: java/Path_Sum_II.java
// Backtracking
// Time is O(n), space is O(n).

public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        LinkedList<Integer> list = new LinkedList<Integer>();
        if (root == null) return result;
        list.add(root.val);
        generate(result, list, root, sum);
        return result;
    }
    public void generate(List<List<Integer>> result, LinkedList<Integer> list, TreeNode root, int sum) {
        if (root.left == null && root.right == null) {
            if (root.val == sum)
                result.add((LinkedList<Integer>)list.clone());
            return;
        }
        if (root.left != null) {
            list.add(root.left.val);
            generate(result, list, root.left, sum - root.val);
            list.remove(list.size()-1);
        }
        if (root.right != null) {
            list.add(root.right.val);
            generate(result, list, root.right, sum - root.val);
            list.remove(list.size()-1);
        }
        return;
    }
}
